aws/sdk/integration-tests/s3/tests/throughput-timeout.rs

```diff
@@ -0,0 +1,256 @@
+    let stream = futures_util::stream::unfold(1, move |state| {
+        let sleep = sleep_clone.clone();
+        async move {
+            if state > 255 {
+                None
+            } else {
+                sleep.sleep(Duration::from_secs(1)).await;
+                Some((
+                    Result::<_, Infallible>::Ok(Bytes::from(vec![state as u8])),
+                    state + 1,
+                ))
+            }
+        }
+    });
+    let body = ByteStream::new(SdkBody::from(hyper::body::Body::wrap_stream(stream)));
+    let body = body.map(move |body| {
+        let time_source = time_clone.clone();
+        // Throw an error if the stream sends less than 1 byte per second
+        let minimum_throughput = (1u64, Duration::from_secs(1));
+        SdkBody::from_dyn(aws_smithy_http::body::BoxBody::new(
+            MinimumThroughputBody::new(time_source, body, minimum_throughput),
+        ))
+    });
+
+    let res = body
+        .collect()
+        .await
+        .expect("no streaming error occurs because data is sent fast enough")
+        .to_vec();
+    assert_eq!(255.0, time_source.seconds_since_unix_epoch());
+    assert_eq!(Duration::from_secs(255), sleep.total_duration());
+    assert_eq!(*EXPECTED_BYTES, res);
+}
+
+#[tokio::test]
+async fn test_throughput_timeout_greater_than() {
+    let _logs = capture_test_logs();
+    let (time_source, sleep) = instant_time_and_sleep(UNIX_EPOCH);
+    let time_clone = time_source.clone();
+    let sleep_clone = sleep.clone();
+
+    // Will send ~1 byte per second.
+    let stream = futures_util::stream::unfold(1, move |state| {
+        let sleep = sleep_clone.clone();
+        async move {
+            if state > 255 {
+                None
+            } else {
+                sleep.sleep(Duration::from_secs(1)).await;
+                Some((
+                    Result::<_, Infallible>::Ok(Bytes::from(vec![state as u8])),
+                    state + 1,
+                ))
+            }
+        }
+    });
+    let body = ByteStream::new(SdkBody::from(hyper::body::Body::wrap_stream(stream)));
+    let body = body.map(move |body| {
+        let time_source = time_clone.clone();
+        // Throw an error if the stream sends less than 1 byte per 2s
+        let minimum_throughput = (1u64, Duration::from_secs(2));
+        SdkBody::from_dyn(aws_smithy_http::body::BoxBody::new(
+            MinimumThroughputBody::new(time_source, body, minimum_throughput),
+        ))
+    });
+    let res = body
+        .collect()
+        .await
+        .expect("no streaming error occurs because data is sent fast enough")
+        .to_vec();
+    assert_eq!(255.0, time_source.seconds_since_unix_epoch());
+    assert_eq!(Duration::from_secs(255), sleep.total_duration());
+    assert_eq!(*EXPECTED_BYTES, res);
+}
+
+// One mebibyte of zeroes.
+const LARGE_BODY: &[u8] = &[0u8; 1024 * 1024];
+
+#[tokio::test]
+async fn test_throughput_timeout_for_upload() {
```

Comment:
This is currently failing due to a hyper error that reminds me of when I was doing the concurrency tests.

```
dispatch failure: io error: error writing a body to connection: Connection reset by peer (os error 54): Connection reset by peer (os error 54) (DispatchFailure(DispatchFailure { source: ConnectorError { kind: Io, source: hyper::Error(BodyWrite, Os { code: 54, kind: ConnectionReset, message: "Connection reset by peer" }), connection: Unknown } }))
```

It's like it's dying when the inner body returns an error. I want to figure this out before merging this PR.